woocommerce/woocommerce
# Refactor BlockTemplatesController

### Changes proposed in this Pull Request:

Part of https://github.com/woocommerce/woocommerce/issues/42251.
Closes https://github.com/woocommerce/woocommerce/issues/44264.
Closes https://github.com/woocommerce/woocommerce/issues/42191.
Closes https://github.com/woocommerce/woocommerce/issues/44453. (**Update:** the testing steps for this issue have been added to the PR description after merge, that's because we didn't notice this PR was fixing #44453 until after merge)

This PR refactors `BlockTemplatesController.php` with the goal to extract all the logic which is specific to certain templates to their own files.

Based on that, every template and template part now has its own class, where the slug, title, description and fallback template are define, as well as any custom logic needed for the template to work. Templates inherit from the `AbstractTemplate` class, which takes care of registering the template into the `BlockTemplatesRegistry`, that's a class that keeps an up-to-date list of registered templates, to make it easy to retrieve data from them (ie: titles, descriptions, etc.).

**Next steps:**

* In each template we still need to hook into the `template_redirect` action. I hope this can be simplified in the future, but given that it involves code from WooCommerce core, it probably makes sense to handle it in https://github.com/woocommerce/woocommerce/issues/44540.
* We have a method called `template_is_eligible_for_product_archive_fallback()`, it might make sense to make it more generic so it allows any template to fall back to any other one. For now, I left it as-is to avoid increasing the complexity of this PR.

### How to test the changes in this Pull Request:

**Make sure there are no regressions related to block templates**

Test it as you wish and feel free to go outside of what's described below. In any case, some ideas of things to test:

_With a classic theme (ie: Storefront)._

1. Make sure there are no regressions when visiting WooCommerce pages. Ie, try going to the _Shop_ page, then visit a single product page, add it to the cart and visit _Cart_ and _Checkout_ pages. Make sure there aren't PHP errors and there isn't anything abnormal.

_With a block theme without custom WooCommerce templates (ie: Twenty Twenty-Four)_

1. Without any template customizations, make sure you can visit the _Shop_ page and the single product page in the frontend without issues.
2. Now, go to Appearance > Editor > Templates and edit the Product Catalog, Single Product and Page: Cart templates and the Mini-Cart template part. You can make a dummy edit like adding a "Hello world from the user-customized template" paragraph.
3. Now, visit the _Shop_ page, the Products by Category page (it should inherit the design from the _Shop_ page), Single Product page and Cart page. Verify the edits are visible in all of them. Add the Mini-Cart to the header of your site or to a new post and verify that when opening the Mini-Cart drawer, the custom template is loaded.
4. Go back to Appearance > Editor > Templates and revert some of the edits you just did (don't revert all). Verify reverting them works and there are no errors.

_With a block theme with custom WooCommerce templates (ie: Tsubaki)_

1. Visit some templates in the frontend and verify the theme templates are visible.
2. Go to Appearance > Editor > Templates and make some edits. When starting to edit the template, verify that the template from the theme is rendered, instead of the template from WooCommerce (note: Cart, Checkout and Order Confirmation don't have specific templates in Tsubaki, that's expected).
3. Save the edits and verify they are visible in the frontend.
4. Make sure edits can be reverted without errors.

**Make sure Mini-Cart template part from the theme is correctly assigned to the Mini-Cart template part area (https://github.com/woocommerce/woocommerce/issues/44264)**

1. With a block theme with a custom Mini-Cart template part (ie: Tsubaki).
2. Go to Appearance > Editor > Patterns (`/wp-admin/site-editor.php?path=%2Fpatterns&categoryType=wp_template_part`)
3. Verify the Mini-Cart template part is in the Mini-Cart template part area.

**Verify correct template is rendered with certain settings (#44453)**

1. With a theme with custom WooCommerce templates (ie: Tsubaki).
2. Go to WooCommerce > Settings > Products > Advanced and disable _Use the product attributes lookup table for catalog filtering_.
3. With a block theme with custom WooCommerce templates (ie: Tsubaki) go to `/shop/?product_cat=Clothing&filter_size=large`. Note: the values of `product_cat` and `filter_xyz` might vary depending on the categories and attributes of your store. In case of doubt, you can import the [products from `sample-data`](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/sample-data).
4. Verify the page is rendered correctly.

Before | After
--- | ---
![image](https://github.com/woocommerce/woocommerce/assets/3616980/12413fda-2c24-49d7-8b21-3cbb4537776f) | ![image](https://github.com/woocommerce/woocommerce/assets/3616980/2ab71af9-754c-46db-be31-9e25598f0167)

### Changelog entry


<details>

#### Significance




#### Type


#### Message

Updated the template logic used by block themes, to make it more performant and resilient

#### Comment 

</details>
